### PR TITLE
Improve inlined line numbers

### PR DESCRIPTION
**Pull request by carnaval**
- start to remove filenames from line number nodes
- get closer to remove the "old" Expr(:line, ...) form
- improve dwarf generation for inline function line numbers
  (get closer to what clang does)

I didn't check for regressions and sysimg size yet. Also the patch needs the frontend improvements from Jeff to remove filenames from the AST entierely (according to him, should be done today).

Probably not working on anything else than LLVM 3.7, I don't feel like fighting my way through the ifdef jungle just yet. Shouldn't we delete everything older than 3.7 now anyway ?

### CODE REVIEW

**Comment by vtjnash** [comment]
> Shouldn't we delete everything older than 3.7 now anyway ?

not until it is equally performant. otherwise, how would you know when you succeed?

**Comment by carnaval** [comment]
It's starting to look ok. It handles arbitrary nested inlined functions, output the correct lambda infos and spec types.

The way it does it is kind of hacky but LLVM's dwarf reader does not make it possible to easily access custom argument so we store some things in the inlined function name.

The only case that does not work is when we do single expression inlining, we'll wait for the linear IR which solves it by definition.

**Comment by vtjnash** [comment]
what's the status of this / what does it still need to be merged?

**Comment by ViralBShah** [comment]
We could delete everything but LLVM 3.3 and 3.7 support.

**Comment by carnaval** [comment]
@Keno alright, it's rebased on top of your cleanup of those horrible files (that was painful, but I agree it's nicer now)

It's still missing a couple things :
- put back support for file name changing in the middle of a function because of macros (the inlining case works of course since that was the whole point)
- fix the build for llvm < 3.7
- use the constant table index of the lambda instead of decompressing the ast for the lookup

The dwarf structure should look ok though, but I only looked at the llvm DI side of things, not at dwarfdump yet. Please do check that it behaves better if you're in the process of parsing the dwarf for the debugger anyway :-)

File: src/disasm.cpp [hunk 1]
```diff
@@ -240,7 +240,7 @@ static int OpInfoLookup(void *DisInfo, uint64_t PC, uint64_t Offset, uint64_t Si
     size_t inlinedat_line;
     jl_lambda_info_t *outer_linfo;
     int fromC;
-    jl_getFunctionInfo(&name, &filename, &line, &inlinedat_file, &inlinedat_line, &outer_linfo, pointer, &fromC, skipC, 1);
+    //jl_getFunctionInfo(&name, &filename, &line, &inlinedat_file, &inlinedat_line, &outer_linfo, pointer, &fromC, skipC, 1);
```
Comment (Keno) [inline src/disasm.cpp]: ?

**Comment by Keno** [comment]
Whitespace in source file is killing CI.

**Comment by JeffBezanson** [comment]
Could these be called `push_loc` and `pop_loc` instead of `push_lambda`? For macro locations we would need to put in a file name instead of a LambdaInfo.

**Comment by JeffBezanson** [comment]
Also, please merge as soon as you can and I'll fix macro line numbers after.

**Comment by carnaval** [comment]
alright this is now on top of jb/linear3 and uses a field in the linfo as we discussed.

still need to fix llvm 3.3 build I think, let's first see what CI says about that

**Comment by carnaval** [comment]
(teaser
```
ERROR: LoadError: hi
 in f(::Int64, ::Bool) at /home/carnaval/dev/julia/repro2.jl:5 [inlined]
 in g(::Int64, ::Bool) at /home/carnaval/dev/julia/repro2.jl:13 [inlined]
 in h(::Bool) at /home/carnaval/dev/julia/repro2.jl:19
```
with all three frames from the same physical function)

**Comment by yuyichao** [comment]
(
> hi

Oscar's favorate English word =D )

This is awesome!

**Comment by vtjnash** [comment]
fixes #13725

**Comment by Keno** [comment]
@JeffBezanson @carnaval Can we get this merged soon?

**Comment by carnaval** [comment]
well it needs jb/linear3 to be merged first, and it's also broken unfortunately. Apart from at least a bug, keeping all those lambdas around is also putting quite a strain on the serializer apparently :/ I'm having a look.

**Comment by JeffBezanson** [comment]
Ok, let's merge jb/linear3 today.

**Comment by JeffBezanson** [comment]
Ok, done.

**Comment by JeffBezanson** [comment]
Any update here?

**Comment by vtjnash** [comment]
we're retiring workers much faster, so it appears that memory usage is a fair bit higher. with the latest change, it looks like we are just barely able to sneak this through CI, but two of the tests need to be updated for the changes.

**Comment by tkelman** [comment]
Note that the 32 bit linux buildbot has been running out of memory for several days, preventing the build of (gallium-compatible) nightlies. Would like to avoid making it any worse.

Will also need some squashing cleanup.

**Comment by carnaval** [comment]
yeah all those temporary commits are just to try and see if the mem usage actually improves as I fix things, will squash.

I'm also planning to investigate more closely why, for example, linalg/triangular needs 1GB of resident memory (even on master) and I don't think generated code is the only culprit. We'll see.

**Comment by vtjnash** [comment]
>Note that the 32 bit linux buildbot has been running out of memory for several days, preventing the build of (gallium-compatible) nightlies. Would like to avoid making it any worse.

would be good to open these observations as issues (most of us don't watch the buildbots too closely). It looks like the last successful build was the merge commit for https://github.com/JuliaLang/julia/pull/15609. the next commits were trivial cleanup, so I'm thinking jb/linear3 seems to be the smoking gun here?

**Comment by tkelman** [comment]
Yes I think that's right. There was one later successful build which threw me off.

**Comment by tkelman** [comment]
I managed to fix the issue for now by rebooting that specific build worker. I think what happened is extra julia processes were accumulating and not getting killed properly when jobs timed out. Don't know whose bug that is, but it's okay for now.

File: src/debuginfo.cpp [hunk 2]
```diff
@@ -1183,34 +1210,83 @@ void jl_getFunctionInfo(char **name, char **filename, size_t *line,
...
+};
+enum section_type {
+    Sect_Code, Sect_Data, Sect_RoData
+};
+static section_memory *section_mem[3];
+#include <sys/mman.h>
+#ifdef _OS_DARWIN_
+#define MAP_ANONYMOUS MAP_ANON
+#endif
+uint8_t *juliaAllocateSection(uintptr_t size, unsigned align, section_type typ)
```
Comment (vtjnash) [inline src/debuginfo.cpp]: this seems unrelated to the PR, can it be pulled out into a separate commit?
